### ISSUE TITLE
Gradient mismatch when using fp16 and freezing part parameters of model

### ISSUE THREAD

**Issue by SefaZeng** [facebookresearch/fairseq#4661]
### Before asking:
1. search the issues.
2. search the docs.

#### What is your question?
I try to freeze part parameters of a model and train with fp16. But the parameter with `required_grad=True` do not update with the optimizer and some other parameters with `required_grad=False` are updating.
#### What's your environment?

 - fairseq Version (e.g., 1.0 or main): main
 - PyTorch Version (e.g., 1.0) 1.10
 - OS (e.g., Linux):
 - How you installed fairseq (`pip`, source): source
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

**Comment by robotsp** [comment]
@SefaZeng  Would you please provide your code? I have the same question.
